robottelo/products.py

```diff
@@ -0,0 +1,674 @@
+        self._distro = value
+
+    @property
+    def content_type(self):   # type: () -> str
+        return self._type
+
+    def __repr__(self):
+        return '<Repo type: {0}, url: {1}, object: {2}>'.format(
+            self.content_type, self.url, hex(id(self)))
+
+
+class YumRepository(BaseRepository):
+    """Custom Yum repository"""
+    _type = REPO_TYPE_YUM  # type: str
+
+
+class DockerRepository(BaseRepository):
+    """Custom Yum repository"""
+    _type = REPO_TYPE_DOCKER  # type: str
+
+
+class PuppetRepository(BaseRepository):
+    """Custom Yum repository"""
+    _type = REPO_TYPE_PUPPET  # type: str
+
+
+class GenericRHRepository(BaseRepository):
+    """Generic RH repository"""
+    _distro = DISTRO_DEFAULT  # type: str
+    _key = None  # type: str
+    _repo_data = None  # type: Dict
+    _url = None  # type: Optional[str]
+
+    def __init__(self, distro=None, key=None, cdn=False, url=None):
+        super(GenericRHRepository, self).__init__()
+
+        if key is not None and self.key:
+            raise Exception('Repository key already defined')
+
+        if key is not None:
+            self._key = key
+
+        if url is not None:
+            self._url = url
+
+        self._cdn = bool(cdn)
+
+        self.distro = distro
+
+    @property
+    def url(self):
+        return self._url
+
+    @property
+    def cdn(self):
+        return bool(self._cdn or settings.cdn or not self.url)
+
+    @property
+    def key(self):
+        return self._key
+
+    @property
+    def distro(self):
+        return self._distro
+
+    @distro.setter
+    def distro(self, distro):  # type: (str) -> None
+        """Set a new distro value, we have to reinitialise the repo data also,
+        if not found raise exception
+        """
+
+        if distro is not None and distro not in DISTROS_SUPPORTED:
+            raise Exception('distro "{0}" not supported'.format(distro))
+        if distro is None:
+            distro = DISTRO_DEFAULT
+        repo_data = self._get_repo_data(distro)
+        if repo_data is None:
+            raise Exception(
+                'Repository data not found for distro {}'.format(distro))
+        self._distro = distro
+        self._repo_data = repo_data
+
+    def _get_repo_data(self, distro=None):  # type: (Optional(str)) -> Dict
+        """Return the repo data as registered in constant module and bound
+        to distro.
+        """
+        if distro is None:
+            distro = self.distro
+        repo_data = None
+        for _, data in REPOS.items():
+            repo_key = data.get('key')
+            repo_distro = data.get('distro')
+            if repo_key == self.key and repo_distro == distro:
+                repo_data = data
+                break
+        return repo_data
+
+    @property
+    def repo_data(self):  # type: () -> Dict
+        if self._repo_data is not None:
+            return self._repo_data
+        self._repo_data = self._get_repo_data()
```

Comment:
Did you consider using something like `cached_property` not to reinvent the wheel for lots of properties? `cached_property` is a requirement for airgun anyways, so you can use it in robottelo as well since it'll be installed.

Reply:
This is already cached in the class line 367-368 , other-wise relaunch the search

Reply:
My suggestion was to replace
```python
    @property
    def repo_data(self):  # type: () -> Dict
        if self._repo_data is not None:
            return self._repo_data
        self._repo_data = self._get_repo_data()
        return self._repo_data
```
with
```python
    @cached_property
    def repo_data(self):  # type: () -> Dict
        return self._get_repo_data()
```
which looks even simpler to me and the value will be cached the same way.
Anyhow, not a blocker, just an idea.